spark/src/main/java/org/apache/iceberg/spark/source/BaseDataReader.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.spark.source;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Iterables;
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.Map;
+import org.apache.iceberg.CombinedScanTask;
+import org.apache.iceberg.FileScanTask;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.encryption.EncryptedFiles;
+import org.apache.iceberg.encryption.EncryptionManager;
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.io.InputFile;
+import org.apache.spark.rdd.InputFileBlockHolder;
+import org.apache.spark.sql.sources.v2.reader.InputPartitionReader;
+
+/**
+ * @param <T> Base class of readers to read data as objects of type @param &lt;T&gt;
+ */
+@SuppressWarnings("checkstyle:VisibilityModifier")
+abstract class BaseDataReader<T> implements InputPartitionReader<T> {
+  final Iterator<FileScanTask> tasks;
+  final Schema tableSchema;
+  final Schema expectedSchema;
+  final FileIO fileIo;
+  final Map<String, InputFile> inputFiles;
+  final boolean caseSensitive;
+
+  Iterator<T> currentIterator;
+  Closeable currentCloseable = null;
+  T current = null;
+  final int batchSize;
```

Comment:
We need to clean up how these instance fields are shared. Some, like `tableSchema`, `expectedSchema`, `caseSensitive` don't need to be held by this class and should be removed. I think we need a `protected` method to get an `InputFile` from `inputFiles`. And I think that `tasks` isn't shared and can be private.

Reply:
I think `batchSize` can also be removed.

Reply:
batchSize is meant really for ColumnarBatch case which will get added shortly.

Reply:
Similarly, the other fields are needed for the BatchTaskDataReader. So makes sense to leave them in the base class.

Reply:
Isn't `batchSize` specific to the batch reader?

For the others, I think the question isn't whether these are used by both classes, but whether it makes sense to have them in the base class. Maybe it does in some cases, but I think if it is just an instance field, there's no need to put it in the base class when it could be a private field in both child classes -- because the cost of maintaining two child fields is small.